### DISCUSSION TITLE
Should handlers be functions only?

### DISCUSSION THREAD

**Discussion by marioidival** [actix/actix-web#2321] [Q&A] [answered]
I'm tying do a sort of handlers in an struct, how I can this with actix-web?

Ex:
```rust
struct A;
impl A {
   async fn foo // omit
}

let a = A();
.route("/foo", web::get().to(a.foo)
```

**Comment by aliemjay** [comment]
You want the handler to store the data of `a`. This is possible by defining a closure that captures `a`:
```rust
.to(|| async move { a.foo().await }) 
```

**Reply by marioidival** [reply]
Sorry about that, @aliemjay , I do not use async because I had a problem with the Responder but my correct example:
```rust
pub struct Controller;
impl Controller {
  pub fn handler(&self) -> impl Responder {
    HttpResponse::Ok().body("some response")
  }
}

// omit
let c = Controller;
// omit
.route("/foo", web::cet().to(c.handler)
```

This not working.

**Reply by marioidival** [reply]
> You want the handler to store the data of `a`. This is possible by defining a closure that captures `a`:
> 
> ```rust
> .to(|| async move { a.foo().await }) 
> ```

This example throw an error
```error[E0525]: expected a closure that implements the `Fn` trait, but this closure only implements `FnOnce`
   --> src/main.rs:152:48
    |
152 | ...                   web::post().to(|| async move { my_controller.handler().await }),
    |                                   -- ^^^^^^^^^^^^^ --------------------------------- closure is `FnOnce` because it moves the variable `my_controller` out of its environment
    |                                   |  |
    |                                   |  this closure implements `FnOnce`, not `Fn`
    |                                   the requirement to implement `Fn` derives from here```

**Comment by aliemjay** [comment, marked as answer]
Sorry, I made a mistake. This should work for async:
```rust
.to(move || a.foo())
```
For non-async:
```rust
.to(move || {
    let response = a.foo();
    async { response }
})
```

**Reply by marioidival** [reply]
```command

error: lifetime may not live long enough
   --> src/main.rs:165:55
    |
165 | ...                   web::get().to(move || async { control.handler() }),
    |                                     ------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
    |                                     |     |
    |                                     |     return type of closure `impl std::future::Future` contains a lifetime `'2`
    |                                     lifetime `'1` represents this closure's body
    |
    = note: closure implements `Fn`, so references to captured variables can't escape the closure

error: aborting due to previous error

error: could not compile `actix`
```

**Reply by marioidival** [reply]
It's strange :/

**Reply by aliemjay** [reply]
The problem is the return type of `control.handler()` is not `'static` (i.e. it contains a non-static reference). This is not allowed in actix-web. You can fix it, for example, by cloning the referenced data or put it under `Rc`.

**Reply by aliemjay** [reply]
You can publish your implementation of  `control.handler()` if you like for me to provide a more proper fix.

**Reply by aliemjay** [reply]
Please see my updated response. There is no problem with your implementation. I should've tested it locally before suggesting it :)

**Reply by marioidival** [reply]
It works!

**Reply by marioidival** [reply]
Now, my problem is how I can pass the HttpRequest into it 🗡️

**Reply by aliemjay** [reply]
This would be
```rust
.to(move |req| {
    let response = controller.handler(req);
    async { response }
})
```

**Reply by aliemjay** [reply]
It works but it may be tricky and not user-friendly solution and you may face more problems with it.

I recommend that you have all your handlers as async functions and you can then access `controller` in them by passing it to `App::app_data`. You can see the docs for app state in https://actix.rs/docs/application/

**Comment by robjtede** [comment]
See that the handler trait includes a `'static` bound.
https://github.com/actix/actix-web/blob/master/src/handler.rs#L22

In effect, it is pretty difficult to use struct methods as handlers in a useful way since you need to move the struct into the closure to use it. This is why all the teaching materials use functions; the framework is designed around it. So yes, handlers should be functions/closures.
